src/linear_algebra/matrix.lean

```diff
@@ -413,6 +413,56 @@ by rw [@linear_equiv.findim_eq R (matrix m n R) _ _ _ _ _ _ (linear_equiv.uncurr
 
 end finite_dimensional
 
+section reindexing
+
+variables {l' m' n' : Type w} [fintype l'] [fintype m'] [fintype n']
+variables {R : Type v}
+
+/-- The natural map that reindexes a matrix's rows and columns with equivalent types is an
+equivalence. -/
+def reindex (eₘ : m ≃ m') (eₙ : n ≃ n') : matrix m n R ≃ matrix m' n' R :=
+{ to_fun    := λ M i j, M (eₘ.symm i) (eₙ.symm j),
+  inv_fun   := λ M i j, M (eₘ i) (eₙ j),
+  left_inv  := λ M, by simp,
+  right_inv := λ M, by simp, }
+
+@[simp] lemma reindex_apply (eₘ : m ≃ m') (eₙ : n ≃ n') (M : matrix m n R) :
+  reindex eₘ eₙ M = λ i j, M (eₘ.symm i) (eₙ.symm j) :=
+rfl
+
+@[simp] lemma reindex_symm_apply (eₘ : m ≃ m') (eₙ : n ≃ n') (M : matrix m' n' R) :
+  (reindex eₘ eₙ).symm M = λ i j, M (eₘ i) (eₙ j) :=
+rfl
+
+/-- The natural map that reindexes a matrix's rows and columns with equivalent types is a linear
+equivalence. -/
+def reindex_linear_equiv [semiring R] (eₘ : m ≃ m') (eₙ : n ≃ n') :
+  matrix m n R ≃ₗ[R] matrix m' n' R :=
+{ map_add'  := λ M N, rfl,
+  map_smul' := λ M N, rfl,
+..(reindex eₘ eₙ)}
+
```

Comment:
I think you need a `@[simp]` lemma (also for the algebra and Lie algebra versions) that says what happens when you apply this equivalence to an argument.

Reply:
Sure, now done!